### PR TITLE
TLS (Issue #43)

### PR DESCRIPTION
**Pull request by Kejmer**
This is a draft of WIP PR for adding TLS connections.

Fixes: https://github.com/scylladb/scylla-rust-driver/issues/43
## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [] All commits compile, pass static checks and pass test.
- [ ] PR description sums up the changes and reasons why they should be introduced.
- [ ] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by psarna** [comment]
I see two main issues:
1. We should not rely on DNS for certificates, and I'm not entirely sure how it's supposed to work at all, given that the code tries to do a DNS lookup on a `SocketAddress`, which is just an IP address. Instead, it should be absolutely fine to just pass the IP address directly.
2. Does `tokio-rustls` allow self-signed certificates? It looks like it does not, which is going to be a problem for testing, where Scylla uses self-signed certificates. Most libraries allow ignoring the fact that the cert is self-signed, but I do not see such option for tokio-rustls (ref: https://github.com/ctz/rustls/issues/124). Without this capability, writing tests is going to become more complex, since a self-signed cert will just result in `Error: IO Error: invalid certificate: CAUsedAsEndEntity` error.

Please test this implementation against a Scylla instance with SSL enabled. Here's what you can add to scylla.yaml configuration file to enable SSL:
```yaml
client_encryption_options:
   enabled: true
   certificate: conf/scylla.crt
   keyfile: conf/scylla.key
   truststore: conf/cadb.pem
   require_client_auth: False
```
, assuming that you already generated the certificate with openssl (with something like `openssl req -x509 -newkey rsa:4096 -keyout scylla.key -out scylla.crt -days 365` or similar).

File: scylla/src/transport/connection.rs [hunk 1]
```diff
@@ -89,12 +100,38 @@ impl Connection {
...
+                let (task, handle) = Self::router(stream, receiver).remote_handle();
+                tokio::task::spawn(task);
+                handle
+            },
+            CustomStream::Tls(stream) => {
+                let (task, handle) = Self::router(stream, receiver).remote_handle();
+                tokio::task::spawn(task);
+                handle
+            },
+        };
```
Comment (piodul) [inline scylla/src/transport/connection.rs:134]: Unless you are going to use the CustomStream enum somewhere else, I think we can get rid of it. You could add a following method to the connection:

```rust

fn run_router(
    stream: (impl AsyncRead + AsyncWrite + Send + 'static),
    receiver: mpsc::Receiver<Task>,
) -> RemoteHandle<()> {
    let (task, handle) = Self::router(stream, receiver).remote_handle();
    tokio::task::spawn(task);
    handle
}
```

And change the above code to:

```rust
let _worker_handle = match config.use_tls {
    true => {
        let mut tls_config = ClientConfig::new();
        tls_config
            .root_store
            .add_server_trust_anchors(&webpki_roots::TLS_SERVER_ROOTS);
        let connector = TlsConnector::from(Arc::new(tls_config));
        let domain = addr.to_string();
        let domain = DNSNameRef::try_from_ascii_str(&domain).map_err(|_| {
            std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid dnsname")
        })?;
        Self::run_router(connector.connect(domain, stream).await?, receiver)
    }
    false => Self::run_router(stream, receiver),
};
```

This gets rid of the Clippy's complaint about the enum variant sizes varying in size too much.

File: scylla/src/transport/connection.rs [hunk 2]
```diff
@@ -1,8 +1,10 @@
 use bytes::Bytes;
 use futures::{future::RemoteHandle, FutureExt};
-use tokio::net::{tcp, TcpSocket, TcpStream};
+use tokio::net::{TcpSocket, TcpStream};
+use tokio::io::{split, AsyncRead, AsyncWrite};
 use tokio::sync::{mpsc, oneshot};
 
+
```
Comment (piodul) [inline scylla/src/transport/connection.rs:13]: Nit: unnecessary whitespace added

File: scylla/src/transport/connection.rs [hunk 3]
```diff
@@ -91,8 +94,21 @@ impl Connection {
...
         let (sender, receiver) = mpsc::channel(128);
 
-        let (fut, _worker_handle) = Self::router(stream, receiver).remote_handle();
-        tokio::task::spawn(fut);
+        let _worker_handle = match config.use_tls {
+            true => {
+                let mut tls_config = ClientConfig::new();
+                tls_config
+                    .root_store
+                    .add_server_trust_anchors(&webpki_roots::TLS_SERVER_ROOTS);
```
Comment (psarna) [inline scylla/src/transport/connection.rs:102]: We must allow the user to provide their own root certificates to trust via configuration options. E.g. rustls' docs describes how custom root certificates can be loaded from a file: https://docs.rs/rustls/0.1.0/rustls/index.html

We also shouldn't use webpki_roots and rely on this crate at all - the users can provide such certificates themselves if they want to. Instead, the users should be able to provide these anchors explicitly if they want them.

**Review by psarna — CHANGES_REQUESTED** [review]

**Comment by Kejmer** [comment]
@psarna 
About DNSNameRef - it doesn't perform any dns lookups, it just validates that a string is a valid hostname. 
When connecting via tls we use information about server's hostname. If we want to connect to a scylla node, how can we get its hostname? We can't use IP address, here is an issue we've found: https://github.com/ctz/rustls/issues/184

**Comment by psarna** [comment]
We'd need to explicitly mark in the documentation that TLS is going to work only on hostnames and also return a meaningful error message if somebody passes an IP address here (something like `scylla-rust-driver only supports TLS for hostnames, not IP addresses`).
It's going to be a little tricky to connect to all nodes if we only receive a hostname of one/few... we can try reverse dns lookup or just only connect to known nodes which were explicitly passed to the session builder by their hostname.

 /cc @piodul

**Comment by psarna** [comment]
and for tests, simply use localhost

**Pull request closed without merging**